File: app/services/user_service.py
"""
Serviço de Usuários
Camada de lógica de negócios para usuários
"""
from typing import List, Optional, Dict, Any
from app.models.user import User
from app.repositories.user_repository import UserRepository


class UserService:
    """Serviço para lógica de negócios relacionada a usuários"""
    
    def __init__(self):
        self.repository = UserRepository()
    
    def create_user(self, data: Dict[str, Any]) -> tuple[Optional[User], Optional[str]]:
        """
        Cria um novo usuário
        Retorna: (usuario, erro)
        """
        # Validações
        if not data.get('name'):
            return None, 'Nome é obrigatório'
        
        if not data.get('email'):
            return None, 'Email é obrigatório'
        
        if not data.get('username'):
            return None, 'Username é obrigatório'
        
        if not data.get('password'):
            return None, 'Senha é obrigatória'
        
        # Verificar se email já existe
        if self.repository.exists_by_email(data['email']):
            return None, 'Email já cadastrado'
        
        # Verificar se username já existe
        if self.repository.exists_by_username(data['username']):
            return None, 'Username já cadastrado'
        
        # Criar usuário
        try:
            user = User(
                name=data['name'],
                email=data['email'],
                username=data['username'],
                phone=data.get('phone', ''),
                is_active=data.get('is_active', True)
            )
            user.set_password(data['password'])
            
            created_user = self.repository.create(user)
            return created_user, None
        except Exception as e:
            return None, f'Erro ao criar usuário: {str(e)}'
    
    def get_user_by_id(self, user_id: int) -> Optional[User]:
        """Busca usuário por ID"""
        return self.repository.find_by_id(user_id)
    
    def get_user_by_email(self, email: str) -> Optional[User]:
        """Busca usuário por email"""
        return self.repository.find_by_email(email)
    
    def get_user_by_username(self, username: str) -> Optional[User]:
        """Busca usuário por username"""
        return self.repository.find_by_username(username)
    
    def get_all_users(self, page: int = 1, per_page: int = 10) -> tuple:
        """
        Retorna todos os usuários com paginação
        Retorna: (lista_usuarios, total_usuarios)
        """
        return self.repository.find_all(page, per_page)
    
    def get_active_users(self) -> List[User]:
        """Retorna todos os usuários ativos"""
        return self.repository.find_active_users()
    
    def update_user(self, user_id: int, data: Dict[str, Any]) -> tuple[Optional[User], Optional[str]]:
        """
        Atualiza um usuário existente
        Retorna: (usuario, erro)
        """
        user = self.repository.find_by_id(user_id)
        if not user:
            return None, 'Usuário não encontrado'
        
        # Validações
        if 'email' in data and data['email'] != user.email:
            if self.repository.exists_by_email(data['email'], exclude_id=user_id):
                return None, 'Email já cadastrado'
        
        if 'username' in data and data['username'] != user.username:
            if self.repository.exists_by_username(data['username'], exclude_id=user_id):
                return None, 'Username já cadastrado'
        
        # Atualizar dados
        try:
            user.update_from_dict(data)
            updated_user = self.repository.update(user)
            return updated_user, None
        except Exception as e:
            return None, f'Erro ao atualizar usuário: {str(e)}'
    
    def delete_user(self, user_id: int) -> tuple[bool, Optional[str]]:
        """
        Remove um usuário
        Retorna: (sucesso, erro)
        """
        user = self.repository.find_by_id(user_id)
        if not user:
            return False, 'Usuário não encontrado'
        
        try:
            self.repository.delete(user)
            return True, None
        except Exception as e:
            return False, f'Erro ao deletar usuário: {str(e)}'
    
    def toggle_user_status(self, user_id: int) -> tuple[Optional[User], Optional[str]]:
        """
        Ativa/Desativa um usuário
        Retorna: (usuario, erro)
        """
        user = self.repository.find_by_id(user_id)
        if not user:
            return None, 'Usuário não encontrado'
        
        try:
            user.is_active = not user.is_active
            updated_user = self.repository.update(user)
            return updated_user, None
        except Exception as e:
            return None, f'Erro ao alterar status: {str(e)}'
    
    def count_users(self) -> int:
        """Retorna o total de usuários cadastrados"""
        return self.repository.count()
    
    def authenticate(self, username_or_email: str, password: str) -> tuple[Optional[User], Optional[str]]:
        """
        Autentica um usuário com username/email e senha
        Retorna: (usuario, erro)
        """
        # Tentar buscar por username ou email
        user = self.repository.find_by_username(username_or_email)
        if not user:
            user = self.repository.find_by_email(username_or_email)
        
        if not user:
            return None, 'Usuário ou senha inválidos'
        
        if not user.is_active:
            return None, 'Usuário inativo. Entre em contato com o administrador.'
        
        if not user.check_password(password):
            return None, 'Usuário ou senha inválidos'
        
        return user, None
    
    def get_user_statistics(self) -> Dict[str, Any]:
        """
        Retorna estatísticas dos usuários para o dashboard
        """
        from app.models.user import db
        from sqlalchemy import func
        from datetime import datetime, timedelta
        
        try:
            # Total de usuários
            total_users = self.repository.count()
            
            # Usuários ativos
            active_users = db.session.query(User).filter(User.is_active == True).count()
            
            # Usuários inativos
            inactive_users = total_users - active_users
            
            # Usuários criados hoje
            today = datetime.utcnow().date()
            users_today = db.session.query(User).filter(
                func.date(User.created_at) == today
            ).count()
            
            # Usuários criados na última semana
            week_ago = datetime.utcnow() - timedelta(days=7)
            users_week = db.session.query(User).filter(
                User.created_at >= week_ago
            ).count()
            
            # Usuários criados no último mês
            month_ago = datetime.utcnow() - timedelta(days=30)
            users_month = db.session.query(User).filter(
                User.created_at >= month_ago
            ).count()
            
            return {
                'total_users': total_users,
                'active_users': active_users,
                'inactive_users': inactive_users,
                'users_today': users_today,
                'users_week': users_week,
                'users_month': users_month,
                'active_percentage': round((active_users / total_users * 100), 1) if total_users > 0 else 0
            }
        except Exception as e:
            return {
                'total_users': 0,
                'active_users': 0,
                'inactive_users': 0,
                'users_today': 0,
                'users_week': 0,
                'users_month': 0,
                'active_percentage': 0
            }
    
    def get_monthly_user_registrations(self, months: int = 6) -> List[Dict[str, Any]]:
        """
        Retorna dados de registros mensais para gráficos
        """
        from app.models.user import db
        from sqlalchemy import func
        from datetime import datetime, timedelta
        import calendar
        
        try:
            # Calcular data inicial (X meses atrás)
            end_date = datetime.utcnow()
            start_date = end_date - timedelta(days=months * 30)
            
            # Query para contar usuários por mês
            monthly_counts = db.session.query(
                func.extract('year', User.created_at).label('year'),
                func.extract('month', User.created_at).label('month'),
                func.count(User.id).label('count')
            ).filter(
                User.created_at >= start_date
            ).group_by(
                func.extract('year', User.created_at),
                func.extract('month', User.created_at)
            ).order_by('year', 'month').all()
            
            # Criar dicionário com os dados
            monthly_data = {}
            for row in monthly_counts:
                key = f"{int(row.year)}-{int(row.month):02d}"
                monthly_data[key] = int(row.count)
            
            # Preencher meses sem dados com 0
            result = []
            current_date = start_date.replace(day=1)
            
            while current_date <= end_date.replace(day=1):
                key = current_date.strftime('%Y-%m')
                month_name = calendar.month_name[current_date.month][:3]  # Jan, Feb, etc.
                
                result.append({
                    'month': month_name,
                    'year': current_date.year,
                    'count': monthly_data.get(key, 0),
                    'full_date': current_date.strftime('%Y-%m')
                })
                
                # Próximo mês
                if current_date.month == 12:
                    current_date = current_date.replace(year=current_date.year + 1, month=1)
                else:
                    current_date = current_date.replace(month=current_date.month + 1)
            
            return result
            
        except Exception as e:
            return []

File: src/main/java/com/torneseumprogramador/desafioiamonolito/services/UserService.java
package com.torneseumprogramador.desafioiamonolito.services;

import com.torneseumprogramador.desafioiamonolito.models.User;
import com.torneseumprogramador.desafioiamonolito.repositories.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.YearMonth;
import java.time.format.DateTimeFormatter;
import java.util.*;

/**
 * Serviço para lógica de negócios relacionada a usuários
 */
@Service
public class UserService {

    @Autowired
    private UserRepository repository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    /**
     * Cria um novo usuário
     * @param data Map com os dados do usuário
     * @return Array com [User criado ou null, mensagem de erro ou null]
     */
    @Transactional
    public Object[] createUser(Map<String, Object> data) {
        try {
            // Validações
            if (!data.containsKey("name") || ((String) data.get("name")).trim().isEmpty()) {
                return new Object[]{null, "Nome é obrigatório"};
            }

            if (!data.containsKey("email") || ((String) data.get("email")).trim().isEmpty()) {
                return new Object[]{null, "Email é obrigatório"};
            }

            if (!data.containsKey("username") || ((String) data.get("username")).trim().isEmpty()) {
                return new Object[]{null, "Username é obrigatório"};
            }

            if (!data.containsKey("password") || ((String) data.get("password")).trim().isEmpty()) {
                return new Object[]{null, "Senha é obrigatória"};
            }

            String email = ((String) data.get("email")).trim();
            String username = ((String) data.get("username")).trim();
            String password = (String) data.get("password");

            // Verificar se email já existe
            if (repository.existsByEmail(email)) {
                return new Object[]{null, "Email já cadastrado"};
            }

            // Verificar se username já existe
            if (repository.existsByUsername(username)) {
                return new Object[]{null, "Username já cadastrado"};
            }

            // Criar usuário
            User user = new User();
            user.setName(((String) data.get("name")).trim());
            user.setEmail(email);
            user.setUsername(username);
            user.setPasswordHash(passwordEncoder.encode(password));
            user.setPhone(data.containsKey("phone") ? ((String) data.get("phone")).trim() : "");
            user.setIsActive(data.containsKey("isActive") ? (Boolean) data.get("isActive") : true);

            User createdUser = repository.save(user);
            return new Object[]{createdUser, null};
        } catch (Exception e) {
            return new Object[]{null, "Erro ao criar usuário: " + e.getMessage()};
        }
    }

    /**
     * Busca usuário por ID
     * @param userId ID do usuário
     * @return Optional contendo o usuário se encontrado
     */
    public Optional<User> getUserById(Long userId) {
        return repository.findById(userId);
    }

    /**
     * Busca usuário por email
     * @param email Email do usuário
     * @return Optional contendo o usuário se encontrado
     */
    public Optional<User> getUserByEmail(String email) {
        return repository.findByEmail(email);
    }

    /**
     * Busca usuário por username
     * @param username Username do usuário
     * @return Optional contendo o usuário se encontrado
     */
    public Optional<User> getUserByUsername(String username) {
        return repository.findByUsername(username);
    }

    /**
     * Retorna todos os usuários com paginação
     * @param page Número da página (1-based)
     * @param perPage Itens por página
     * @return Array com [Lista de usuários, Total de usuários]
     */
    public Object[] getAllUsers(int page, int perPage) {
        Pageable pageable = PageRequest.of(page - 1, perPage, Sort.by(Sort.Direction.DESC, "createdAt"));
        Page<User> userPage = repository.findAll(pageable);
        return new Object[]{userPage.getContent(), userPage.getTotalElements()};
    }

    /**
     * Retorna todos os usuários ativos
     * @return Lista de usuários ativos
     */
    public List<User> getActiveUsers() {
        return repository.findByIsActiveTrueOrderByCreatedAtDesc();
    }

    /**
     * Atualiza um usuário existente
     * @param userId ID do usuário
     * @param data Map com os dados a atualizar
     * @return Array com [User atualizado ou null, mensagem de erro ou null]
     */
    @Transactional
    public Object[] updateUser(Long userId, Map<String, Object> data) {
        Optional<User> optionalUser = repository.findById(userId);
        if (!optionalUser.isPresent()) {
            return new Object[]{null, "Usuário não encontrado"};
        }

        User user = optionalUser.get();

        try {
            // Validar email se foi alterado
            if (data.containsKey("email")) {
                String newEmail = ((String) data.get("email")).trim();
                if (!newEmail.equals(user.getEmail())) {
                    if (repository.existsByEmailExcludingId(newEmail, userId)) {
                        return new Object[]{null, "Email já cadastrado"};
                    }
                    user.setEmail(newEmail);
                }
            }

            // Validar username se foi alterado
            if (data.containsKey("username")) {
                String newUsername = ((String) data.get("username")).trim();
                if (!newUsername.equals(user.getUsername())) {
                    if (repository.existsByUsernameExcludingId(newUsername, userId)) {
                        return new Object[]{null, "Username já cadastrado"};
                    }
                    user.setUsername(newUsername);
                }
            }

            // Atualizar outros campos
            if (data.containsKey("name")) {
                user.setName(((String) data.get("name")).trim());
            }
            if (data.containsKey("phone")) {
                user.setPhone(((String) data.get("phone")).trim());
            }
            if (data.containsKey("isActive")) {
                user.setIsActive((Boolean) data.get("isActive"));
            }
            if (data.containsKey("password") && !((String) data.get("password")).trim().isEmpty()) {
                user.setPasswordHash(passwordEncoder.encode((String) data.get("password")));
            }

            User updatedUser = repository.save(user);
            return new Object[]{updatedUser, null};
        } catch (Exception e) {
            return new Object[]{null, "Erro ao atualizar usuário: " + e.getMessage()};
        }
    }

    /**
     * Remove um usuário
     * @param userId ID do usuário
     * @return Array com [sucesso (boolean), mensagem de erro ou null]
     */
    @Transactional
    public Object[] deleteUser(Long userId) {
        Optional<User> optionalUser = repository.findById(userId);
        if (!optionalUser.isPresent()) {
            return new Object[]{false, "Usuário não encontrado"};
        }

        try {
            repository.delete(optionalUser.get());
            return new Object[]{true, null};
        } catch (Exception e) {
            return new Object[]{false, "Erro ao deletar usuário: " + e.getMessage()};
        }
    }

    /**
     * Ativa/Desativa um usuário
     * @param userId ID do usuário
     * @return Array com [User atualizado ou null, mensagem de erro ou null]
     */
    @Transactional
    public Object[] toggleUserStatus(Long userId) {
        Optional<User> optionalUser = repository.findById(userId);
        if (!optionalUser.isPresent()) {
            return new Object[]{null, "Usuário não encontrado"};
        }

        try {
            User user = optionalUser.get();
            user.setIsActive(!user.getIsActive());
            User updatedUser = repository.save(user);
            return new Object[]{updatedUser, null};
        } catch (Exception e) {
            return new Object[]{null, "Erro ao alterar status: " + e.getMessage()};
        }
    }

    /**
     * Retorna o total de usuários cadastrados
     * @return Número total de usuários
     */
    public long countUsers() {
        return repository.count();
    }

    /**
     * Autentica um usuário com username/email e senha
     * @param usernameOrEmail Username ou email
     * @param password Senha
     * @return Array com [User autenticado ou null, mensagem de erro ou null]
     */
    public Object[] authenticate(String usernameOrEmail, String password) {
        // Tentar buscar por username ou email
        Optional<User> optionalUser = repository.findByUsername(usernameOrEmail);
        if (!optionalUser.isPresent()) {
            optionalUser = repository.findByEmail(usernameOrEmail);
        }

        if (!optionalUser.isPresent()) {
            return new Object[]{null, "Usuário ou senha inválidos"};
        }

        User user = optionalUser.get();

        if (!user.getIsActive()) {
            return new Object[]{null, "Usuário inativo. Entre em contato com o administrador."};
        }

        if (!passwordEncoder.matches(password, user.getPasswordHash())) {
            return new Object[]{null, "Usuário ou senha inválidos"};
        }

        return new Object[]{user, null};
    }

    /**
     * Retorna estatísticas dos usuários para o dashboard
     * @return Map com as estatísticas
     */
    public Map<String, Object> getUserStatistics() {
        try {
            long totalUsers = repository.count();
            long activeUsers = repository.countByIsActiveTrue();
            long inactiveUsers = totalUsers - activeUsers;

            LocalDateTime today = LocalDate.now().atStartOfDay();
            LocalDateTime tomorrow = today.plusDays(1);
            long usersToday = repository.countByCreatedAtBetween(today, tomorrow);

            LocalDateTime weekAgo = LocalDateTime.now().minusDays(7);
            long usersWeek = repository.countByCreatedAtAfter(weekAgo);

            LocalDateTime monthAgo = LocalDateTime.now().minusDays(30);
            long usersMonth = repository.countByCreatedAtAfter(monthAgo);

            double activePercentage = totalUsers > 0 ? Math.round((double) activeUsers / totalUsers * 1000.0) / 10.0 : 0.0;

            Map<String, Object> stats = new HashMap<>();
            stats.put("total_users", totalUsers);
            stats.put("active_users", activeUsers);
            stats.put("inactive_users", inactiveUsers);
            stats.put("users_today", usersToday);
            stats.put("users_week", usersWeek);
            stats.put("users_month", usersMonth);
            stats.put("active_percentage", activePercentage);

            return stats;
        } catch (Exception e) {
            Map<String, Object> emptyStats = new HashMap<>();
            emptyStats.put("total_users", 0);
            emptyStats.put("active_users", 0);
            emptyStats.put("inactive_users", 0);
            emptyStats.put("users_today", 0);
            emptyStats.put("users_week", 0);
            emptyStats.put("users_month", 0);
            emptyStats.put("active_percentage", 0.0);
            return emptyStats;
        }
    }

    /**
     * Retorna dados de registros mensais para gráficos
     * @param months Número de meses a retornar
     * @return Lista de mapas com os dados mensais
     */
    public List<Map<String, Object>> getMonthlyUserRegistrations(int months) {
        try {
            List<Map<String, Object>> result = new ArrayList<>();
            LocalDate endDate = LocalDate.now();
            LocalDate startDate = endDate.minusMonths(months);

            YearMonth currentMonth = YearMonth.from(startDate.withDayOfMonth(1));
            YearMonth lastMonth = YearMonth.from(endDate);

            while (!currentMonth.isAfter(lastMonth)) {
                LocalDateTime monthStart = currentMonth.atDay(1).atStartOfDay();
                LocalDateTime monthEnd = currentMonth.atEndOfMonth().atTime(23, 59, 59);

                long count = repository.countByCreatedAtBetween(monthStart, monthEnd);

                Map<String, Object> monthData = new HashMap<>();
                monthData.put("month", currentMonth.getMonth().toString().substring(0, 3));
                monthData.put("year", currentMonth.getYear());
                monthData.put("count", count);
                monthData.put("full_date", currentMonth.format(DateTimeFormatter.ofPattern("yyyy-MM")));

                result.add(monthData);
                currentMonth = currentMonth.plusMonths(1);
            }

            return result;
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }
}
